pkg/k8s/watchers/crd.go

```diff
@@ -0,0 +1,294 @@
+//     "customresourcedefinitions.apiextensions.k8s.io
+//     "customresourcedefinitions" not found".
+//   - If the namespace was set, the following error was observed:
+//     "an empty namespace may not be set when a resource name is provided".
+//
+// The namespace problem can be worked around by using NamespaceIfScoped, but
+// it's been omitted entirely here because it's equivalent in functionality.
+func newListWatchFromClient(c cache.Getter, fieldSelector fields.Selector) *cache.ListWatch {
+	optionsModifier := func(options *metav1.ListOptions) {
+		options.FieldSelector = fieldSelector.String()
+	}
+
+	listFunc := func(options metav1.ListOptions) (runtime.Object, error) {
+		// This lister  will retrieve the CRDs as a metav1.Table object. This
+		// is the same way that `kubectl get crds` returns them. In the
+		// metav1.Table object, it contains cells, which are
+		// metav1.PartialObjectMetadata objects containing the minimal
+		// representation of an objects in K8s (in this case a CRD). Once the
+		// metav1.Table is fetched, it is converted into a
+		// metav1.PartialObjectMetadataList which is what the controller
+		// (informer) expects the object type to be. If we returned the
+		// metav1.Table itself, the controller doesn't know how to handle that
+		// object type because it is not a list type.
+
+		optionsModifier(&options)
+
+		t := &metav1.Table{}
+		err := c.Get().
+			SetHeader("Accept", tableHeader).
+			VersionedParams(&options, metav1.ParameterCodec).
+			Do(context.TODO()).
+			Into(t)
+		if err != nil {
+			return nil, err
+		}
+		return tableToPomList(t), nil
+	}
+	watchFunc := func(options metav1.ListOptions) (watch.Interface, error) {
+		// This watcher will retrieve each CRD that the lister has listed as
+		// individual metav1.PartialObjectMetadata because it is requesting the
+		// apiserver to return objects as such via the "Accept" header.
+
+		optionsModifier(&options)
+
+		options.Watch = true
+		return c.Get().
+			SetHeader("Accept", partialObjHeader).
+			VersionedParams(&options, metav1.ParameterCodec).
+			Watch(context.TODO())
+	}
+	return &cache.ListWatch{ListFunc: listFunc, WatchFunc: watchFunc}
+}
+
+// tableToPomList converts a metav1.Table's cells into individual
+// metav1.PartialObjectMetadata objects, and returns them all placed in a
+// metav1.PartialObjectMetadataList.
+func tableToPomList(t *metav1.Table) *metav1.PartialObjectMetadataList {
+	list := &metav1.PartialObjectMetadataList{
+		TypeMeta: t.TypeMeta,
+		ListMeta: t.ListMeta,
+		Items:    make([]metav1.PartialObjectMetadata, 0, len(t.Rows)),
+	}
+
+	for _, row := range t.Rows {
+		var pom metav1.PartialObjectMetadata
+		if err := json.Unmarshal(row.Object.Raw, &pom); err != nil {
+			// TODO(christarazi): Not sure what to do here. We expected
+			// PartialObjectMetadata.
+			fmt.Printf("chris received %T, expected *metav1.PartialObjectMetadata raw %s\n",
+				row.Object.Object, string(row.Object.Raw))
+			continue
```

Comment:
Maybe logging an info msg? I don't think we want to halt in any way if we encounter an error here.

Reply:
Yes, I think we should log an `Error` because if we get something that we aren't expecting Cilium will wait for the CRDs forever.